tests/e2e/reports/report-date-filter.wdio-spec.js

```diff
@@ -1,108 +1,63 @@
-      },
-      from: '+555',
-      hidden_fields: []
-    },
+    reportFactory.build(
+      { form: 'P', reported_date: moment([2016, 4, 16, 0, 30]).valueOf() },
+      { patient, submitter: user.contact, fields: { lmp_date: 'Feb 15, 2016' },
+      }),
     // one visit half an hour after the end date
-    {
-      fields: {
-        ok: 'Yes!'
-      },
-      form: 'V',
-      type: 'data_record',
-      content_type: 'xml',
-      reported_date: moment([2016, 4, 18, 0, 30]).valueOf(), // month is 0 based in this context
-      contact: {
-        name: 'Sharon',
-        phone: '+555',
-        type: 'person',
-        _id: '3305E3D0-2970-7B0E-AB97-C3239CD22D32',
-        _rev: '1-fb7fbda241dbf6c2239485c655818a69'
-      },
-      from: '+555',
-      hidden_fields: []
-    },
+    reportFactory.build(
+      { form: 'V', reported_date: moment([2016, 4, 18, 0, 30]).valueOf() },
+      { patient, submitter: user.contact, fields: { ok: 'Yes!' },
+      }),
     // one visit half an hour before the end date
-    {
-      fields: {
-        ok: 'Yes!'
-      },
-      form: 'V',
-      type: 'data_record',
-      content_type: 'xml',
-      reported_date: moment([2016, 4, 17, 23, 30]).valueOf(), // month is 0 based in this context
-      contact: {
-        name: 'Sharon',
-        phone: '+555',
-        type: 'person',
-        _id: '3305E3D0-2970-7B0E-AB97-C3239CD22D32',
-        _rev: '1-fb7fbda241dbf6c2239485c655818a69'
-      },
-      from: '+555',
-      hidden_fields: []
-    },
+    reportFactory.build(
+      { form: 'V', reported_date: moment([2016, 4, 17, 23, 30]).valueOf() },
+      { patient, submitter: user.contact, fields: { ok: 'Yes!' },
+      }),
   ];
 
-  const savedUuids = [];
   beforeEach(async () => {
-    await loginPage.cookieLogin();
-    const results = await utils.saveDocs(reports);
-    results.forEach(result => savedUuids.push(result.id));
-  });
+    const results = await utils.saveDocs([ parent, patient, ...reports ]);
+    results.splice(0, 2); // Keeping only reports
+    savedReports = results.map(result => result.id);
 
-  it('by date', async () => {
+    await utils.createUsers([ user ]);
+    await loginPage.login(user);
     await commonElements.goToReports();
-    await (await reportsTab.firstReport()).waitForDisplayed();
+  });
 
-    await reportsTab.filterByDate(moment('05/16/2016', 'MM/DD/YYYY'), moment('05/17/2016', 'MM/DD/YYYY'));
+  it('should filter by date', async () => {
     await (await reportsTab.firstReport()).waitForDisplayed();
 
-    expect(await (await reportsTab.allReports()).length).to.equal(2);
-    expect(await (await reportsTab.reportsByUUID(savedUuids[1])).length).to.equal(1);
-    expect(await (await reportsTab.reportsByUUID(savedUuids[3])).length).to.equal(1);
+    await reportsTab.filterByDate(moment('05/16/2016', 'MM/DD/YYYY'), moment('05/17/2016', 'MM/DD/YYYY'));
+    await browser.pause(300); // Wait for filter to finish
```

Comment:
I think that we should avoid explicit waits as much as we can. If you replace the  `await browser.pause(300)` for `await commonElements.waitForPageLoaded()` it will wait until the filtered reports are displayed.

Reply:
Quite interesting function with nested loops (do-while, for), assuming that we have few elements with `.container-fluid .loader` and that hopefully wdio query aren't expensive then it should be okay. Done!